api/src/main/java/org/pmiops/workbench/compliancetraining/ComplianceTrainingServiceImpl.java

```diff
@@ -239,27 +239,35 @@ public DbUser syncComplianceTrainingStatusAbsorb(DbUser dbUser, Agent agent) thr
       var maybeEnrollment =
           enrollments.stream().filter(e -> e.courseId.equals(courseId)).findFirst();
 
-      if (maybeEnrollment.isEmpty()) {
-        log.severe(
-            String.format(
-                "User `%s` is not enrolled in Absorb course `%s` for access module `%s`. Users are expected to be automatically enrolled in all courses upon visiting Absorb.",
-                dbUser.getUsername(), courseId, accessModuleName));
-        throw new NotFoundException(
-            String.format(
-                "User %s is not enrolled in Absorb course %s", dbUser.getUsername(), courseId));
-      }
-      var enrollment = maybeEnrollment.get();
-
-      // The course is incomplete, do not update the user access module
-      if (enrollment.completionTime == null) {
-        continue;
-      }
+      maybeEnrollment.ifPresentOrElse(
```

Comment:
take it or leave it..but to make things easier to test, I'd probably push things that call out to absorb outside this method, and make `enrollments` an input argument.

With that, this function will be mostly pure, and u'll be able to test out functionality much easier that way

Reply:
I was thinking the same thing (maybe we read the same chapter of a certain book 😉), but I wasn't sure how to handle control flow involving `absorbService.userHasLoggedIntoAbsorb` and `absorbService.getActiveEnrollmentsForUser`. The latter should only be called if the former is true, otherwise there are no enrollments to obtain. It's also not clear to me how to handle the two lines that modify the DB (a side effect) in an FP-friendly way.

I'm going to leave it as-is for now, but I think this function is a perfect level of complexity that refactoring it would make a fantastic FP-focused mobbing session.

Side note: More broadly, this codebase has a lot of instances of new practices being introduced (Kotlin, lodash FP, etc.) without following through on conversions. This has lead to a fractured codebase that I find very complex. For the sake of future new engineers on the team, especially those unfamiliar with FP (like many new grads), I'm hesitant to start introducing FP-isms (even if that's a much lighter lift than a new technology) in a codebase that is primarily imperative without broader agreement or ways to enforce / encourage it. My ask, if you want to move our codebase toward FP, is that you find a way to obtain that buy-in. That doesn't need to be huge, even small agreements like "we shouldn't call APIs in services, only controllers (for testability's sake)" in an eng sync would be enough for me to change.

Reply:
> I'm hesitant to start introducing FP-isms (even if that's a much lighter lift than a new technology) in a codebase that is primarily imperative without broader agreement or ways to enforce / encourage it

Just to be clear, I'm not trying to move things to FPish. I think moving to "output-based" style of unit testing is sth we can strive to in this code base potentially (and I got the sense you'd agree as well). Also reason why I said "take it or leave it" is becuz I don't know this code base well enough to judge whether the suggested approach is feasible, and what I want to achieve for the most part is to nudge you to think in that direction. The thought exercise alone is worthwhile IMO.

Reply:
Thanks, I feel like we're on the same page!